src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,224 @@
+                            params_.data_shape[0])};
+        }
+        cv::Mat tmp = decode_image(mat, mode, ins);
+        if (params_.resize != -1) {
+            resize(tmp, tmp);
+        }
+        if (params_.to_rgb && tmp.channels() != 1) {
+            cv::cvtColor(tmp, tmp, cv::COLOR_BGR2RGB);
+        }
+        cv::Mat dst{
+            params_.data_shape[1], params_.data_shape[2], type, ptr_img_vec};
+        crop(tmp, dst);
+        ptr_img_vec += instance_offset;
+    }
+
+    auto ptr = wrap_cpu_array<data_type::uint8>(std::move(img_vec));
+    auto tsr = make_intrusive<dense_tensor>(
+        size_vector{batch_size,
+                    static_cast<std::size_t>(params_.data_shape[1]),
+                    static_cast<std::size_t>(params_.data_shape[2]),
+                    static_cast<std::size_t>(params_.data_shape[0])},
+        std::move(ptr));
+    return tsr;
+}
+
+cv::Mat
+image_reader::decode_image(const cv::Mat &buf,
+                           int mode,
+                           const instance &ins) const
+{
+    cv::Mat decoded_img = cv::imdecode(buf, mode);
+    if (decoded_img.empty()) {
+        throw std::runtime_error{fmt::format("image decoding failed. Image: ",
+                                             ins.get_data_store().id())};
+    }
+    if (mode == cv::ImreadModes::IMREAD_UNCHANGED &&
+        decoded_img.channels() != 4) {
+        throw std::invalid_argument{
+            fmt::format("Invalid image: {0:s}. Expected 4 channels. Got {1:n}",
+                        ins.get_data_store().id(),
+                        decoded_img.channels())};
+    }
+    return decoded_img;
+}
+
+void
+image_reader::resize(cv::Mat &src, cv::Mat &dst) const
+{
+    if (params_.resize <= 0) {
+        throw std::invalid_argument{fmt::format(
+            "Invalid value for the resize operation: {0:n}. Value should be "
+            "greater than 0",
+            params_.resize)};
+    }
+
+    int new_height, new_width;
+    if (src.rows > src.cols) {
+        new_height = params_.resize * src.rows / src.cols;
+        new_width = params_.resize;
+    }
+    else {
+        new_height = params_.resize;
+        new_width = params_.resize * src.cols / src.rows;
+    }
+    cv::resize(src, dst, cv::Size(new_width, new_height), 0, 0);
+}
+
+void
+image_reader::crop(cv::Mat &src, cv::Mat &dst) const
+{
+    if (src.rows < params_.data_shape[1] || src.cols < params_.data_shape[2]) {
+        throw std::invalid_argument{fmt::format(
+            "Input image [rows: {0:n}, cols: {1:n}] is smaller than "
+            "the output image data_shape [rows: {2:n}, cols: {3:n}]",
+            src.rows,
+            src.cols,
+            params_.data_shape[1],
+            params_.data_shape[2])};
+    }
+
+    // center crop
+    int y = (src.rows - params_.data_shape[1]) / 2;
+    int x = (src.cols - params_.data_shape[2]) / 2;
+
+    cv::Rect roi(x, y, params_.data_shape[2], params_.data_shape[1]);
+    src(roi).copyTo(dst);
```

Comment:
Like other places, this operation can fail, and we do not check it.

Reply:
this method does not return any error code: https://docs.opencv.org/4.2.0/d3/d63/classcv_1_1Mat.html#a33fd5d125b4c302b0c9aa86980791a77